warehouse/oidc/utils.py

```diff
@@ -40,54 +47,84 @@ def find_publisher_by_issuer(session, issuer_url, signed_claims, *, pending=Fals
+    workflow_prefix = f"{repository}/.github/workflows/"
+    workflow_ref = signed_claims["job_workflow_ref"].removeprefix(workflow_prefix)
+
+    publisher_cls = GitHubPublisher if not pending else PendingGitHubPublisher
+
+    publisher = None
+    # If an environment exists in the claim set, try finding a publisher
+    # that matches the provided environment first.
+    if environment := signed_claims.get("environment"):
+        publisher = (
+            session.query(publisher_cls)
+            .filter_by(
+                repository_name=repository_name,
+                repository_owner=repository_owner,
+                repository_owner_id=signed_claims["repository_owner_id"],
+                environment=environment,
+            )
+            .filter(
+                literal(workflow_ref).like(
+                    func.concat(publisher_cls.workflow_filename, "%")
                 )
-                .one_or_none()
             )
+            .one_or_none()
+        )
 
-        # There are no publishers for that specific environment, try finding a
-        # publisher without a restriction on the environment
-        if not publisher:
-            publisher = (
-                session.query(publisher_cls)
-                .filter_by(
-                    repository_name=repository_name,
-                    repository_owner=repository_owner,
-                    repository_owner_id=signed_claims["repository_owner_id"],
-                    environment=None,
-                )
-                .filter(
-                    literal(workflow_ref).like(
-                        func.concat(publisher_cls.workflow_filename, "%")
-                    )
+    # There are no publishers for that specific environment, try finding a
+    # publisher without a restriction on the environment
+    if not publisher:
+        publisher = (
+            session.query(publisher_cls)
+            .filter_by(
+                repository_name=repository_name,
+                repository_owner=repository_owner,
+                repository_owner_id=signed_claims["repository_owner_id"],
+                environment=None,
+            )
+            .filter(
+                literal(workflow_ref).like(
+                    func.concat(publisher_cls.workflow_filename, "%")
                 )
-                .one_or_none()
             )
+            .one_or_none()
+        )
+
+    return publisher
+
+
+def _find_google_publisher(session, signed_claims, *, pending=False):
+    publisher_cls = GooglePublisher if not pending else PendingGooglePublisher
 
+    # The `sub` claim should always be present in the claim set, but is
+    # optional on the registered publisher side. We first check for a
+    # publisher that has both, and then for one with no `sub`.
+    # This has to happen as separate queries to prevent duplicate
+    # results: both variants can be registered, and we always want
+    # to return the more specialized one if possible.
```

Comment:
Flagging for review: we have to do this double-query dance for the same reason that GitHub publishers need it (an optional, unique claim, in this case `sub`).

Since this is a brand new publisher rather than one we're augmenting, we could instead preempt this problem: we could make the unique constraint on `GooglePublisher`s _just_ on the `email` claim, meaning that it wouldn't be possible to have two unique publishers that only vary on `sub`.  This simplifies the code here, at the cost of _potentially_ (I'm not sure how Google's `sub` claim maps to identity) excluding some valid unique combinations of `(email, sub)`.

Thoughts @di?

Reply:
I _think_ the only time we'd have the same `email` with a different `sub` claim is if a service account was deleted & recreated, and a new publisher was added. That probably means that the old `sub` claim will never occur after that point, so this would probably be fine, however I think the UX here would be a little confusing, because a user would consider these separate publishers. So I think the double-query is an OK tradeoff for that.

Reply:
Makes sense! Resolving.